cmd/swarm/swarm-smoke/upload_and_sync.go

```diff
@@ -17,84 +17,109 @@
 package main
 
 import (
-	"crypto/md5"
-	crand "crypto/rand"
+	"bytes"
 	"fmt"
-	"io"
 	"math/rand"
 	"sync"
 	"time"
 
 	"github.com/ethereum/go-ethereum/log"
 	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethereum/go-ethereum/swarm/testutil"
 	"github.com/pborman/uuid"
 
 	cli "gopkg.in/urfave/cli.v1"
 )
 
-func uploadAndSync(c *cli.Context) error {
-	generateEndpoints(scheme, cluster, appName, from, to)
-	seed := int(time.Now().UnixNano() / 1e6)
+func uploadAndSync(ctx *cli.Context, tuid string) error {
+	randomBytes := testutil.RandomBytes(seed, filesize*1000)
 
-	log.Info("uploading to "+endpoints[0]+" and syncing", "seed", seed)
+	errc := make(chan error)
 
-	h := md5.New()
-	r := io.TeeReader(io.LimitReader(crand.Reader, int64(filesize*1000)), h)
+	go func() {
+		errc <- uas(ctx, randomBytes, tuid)
+	}()
+
+	select {
+	case err := <-errc:
+		if err != nil {
+			metrics.GetOrRegisterCounter(fmt.Sprintf("%s.fail", commandName), nil).Inc(1)
+		}
+		return err
+	case <-time.After(time.Duration(timeout) * time.Second):
+		metrics.GetOrRegisterCounter(fmt.Sprintf("%s.timeout", commandName), nil).Inc(1)
+
+		// trigger debug functionality on randomBytes
+
+		return fmt.Errorf("timeout after %v sec", timeout)
+	}
+}
+
+func uas(c *cli.Context, randomBytes []byte, tuid string) error {
+	log.Info("uploading to "+httpEndpoint(hosts[0])+" and syncing", "tuid", tuid, "seed", seed)
 
 	t1 := time.Now()
-	hash, err := upload(r, filesize*1000, endpoints[0])
+	hash, err := upload(randomBytes, httpEndpoint(hosts[0]))
 	if err != nil {
 		log.Error(err.Error())
 		return err
 	}
-	metrics.GetOrRegisterResettingTimer("upload-and-sync.upload-time", nil).UpdateSince(t1)
+	t2 := time.Since(t1)
+	metrics.GetOrRegisterResettingTimer("upload-and-sync.upload-time", nil).Update(t2)
 
-	fhash := h.Sum(nil)
+	fhash, err := digest(bytes.NewReader(randomBytes))
```

Comment:
Do you want to be safe here and generate a control hash from the bytes themselves rather than getting it from the upload?
Or what is the reason to change this

Reply:
I am reverting this, because the previous version is using a random reader, from which we cannot read multiple times.

This was the original version, but was then changed to a reader in order to use less memory.

Reply:
Please use io.TeeReader to read it multiple times. If you think that optimizing memory consumption is not relevant, ignore my comments freely.

Reply:
Please use io.TeeReader to read it multiple times. Was there a problem with the version with it? If you think that memory consumption is not relevant, ignore my comment.

Reply:
I was under the impression I cannot read the same random bytes multiple times. I will review again, and revert if I was wrong.

Reply:
Memory consumption is not really relevant here, but it'd be nice to not be wasteful. Still I don't understand what you mean - we must buffer `randomBytes` so that if the smoke test fails, we extract all references from the actual content.

`bytes.NewReader(randomBytes)` is just creating a reader from the byte slice that we must keep in memory. If we don't keep it in memory, we only have the hash of the content, which is of no use, once the upload-and-sync fails.

Reply:
Ok, if randomBytes are necessary for the core functionality.

Reply:
One question, what do we do with randomBytes when upload-and-sync fails? I see that it is generated in uploadAndSync and then just passed to uas, nothing more. Function uas is only uploading it and generating a hash.

Shouldn't this work the same but without allocating the whole file:

```diff
diff --git a/cmd/swarm/swarm-smoke/upload_and_sync.go b/cmd/swarm/swarm-smoke/upload_and_sync.go
index e129802ed..725ab1ed4 100644
--- a/cmd/swarm/swarm-smoke/upload_and_sync.go
+++ b/cmd/swarm/swarm-smoke/upload_and_sync.go
@@ -17,27 +17,26 @@
 package main
 
 import (
-       "bytes"
+       "crypto/md5"
+       crand "crypto/rand"
        "fmt"
+       "io"
        "math/rand"
        "sync"
        "time"
 
        "github.com/ethereum/go-ethereum/log"
        "github.com/ethereum/go-ethereum/metrics"
-       "github.com/ethereum/go-ethereum/swarm/testutil"
        "github.com/pborman/uuid"
 
        cli "gopkg.in/urfave/cli.v1"
 )
 
 func uploadAndSync(ctx *cli.Context, tuid string) error {
-       randomBytes := testutil.RandomBytes(seed, filesize*1000)
-
        errc := make(chan error)
 
        go func() {
-               errc <- uas(ctx, randomBytes, tuid)
+               errc <- uas(ctx, tuid)
        }()
 
        select {
@@ -55,11 +54,15 @@ func uploadAndSync(ctx *cli.Context, tuid string) error {
        }
 }
 
-func uas(c *cli.Context, randomBytes []byte, tuid string) error {
+func uas(c *cli.Context, tuid string) error {
        log.Info("uploading to "+httpEndpoint(hosts[0])+" and syncing", "tuid", tuid, "seed", seed)
 
+       h := md5.New()
+       size := int64(filesize * 1000)
+       r := io.TeeReader(io.LimitReader(crand.Reader, size), h)
+
        t1 := time.Now()
-       hash, err := upload(randomBytes, httpEndpoint(hosts[0]))
+       hash, err := uploadReader(r, size, httpEndpoint(hosts[0]))
        if err != nil {
                log.Error(err.Error())
                return err
@@ -67,11 +70,7 @@ func uas(c *cli.Context, randomBytes []byte, tuid string) error {
        t2 := time.Since(t1)
        metrics.GetOrRegisterResettingTimer("upload-and-sync.upload-time", nil).Update(t2)
 
-       fhash, err := digest(bytes.NewReader(randomBytes))
-       if err != nil {
-               log.Error(err.Error())
-               return err
-       }
+       fhash := h.Sum(nil)
 
        log.Info("uploaded successfully", "tuid", tuid, "hash", hash, "took", t2, "digest", fmt.Sprintf("%x", fhash))
 
diff --git a/cmd/swarm/swarm-smoke/util.go b/cmd/swarm/swarm-smoke/util.go
index 87abb44b0..4768d44a7 100644
--- a/cmd/swarm/swarm-smoke/util.go
+++ b/cmd/swarm/swarm-smoke/util.go
@@ -213,6 +213,21 @@ func upload(data []byte, endpoint string) (string, error) {
        return swarm.Upload(f, "", false)
 }
 
+func uploadReader(r io.Reader, size int64, endpoint string) (string, error) {
+       swarm := client.NewClient(endpoint)
+       f := &client.File{
+               ReadCloser: ioutil.NopCloser(r),
+               ManifestEntry: api.ManifestEntry{
+                       ContentType: "text/plain",
+                       Mode:        0660,
+                       Size:        size,
+               },
+       }
+
+       // upload data to bzz:// and retrieve the content-addressed manifest hash, hex-encoded.
+       return swarm.Upload(f, "", false)
+}
+
 func digest(r io.Reader) ([]byte, error) {
        h := md5.New()
        _, err := io.Copy(h, r)
```

Reply:
@janos the debug functionality is not implemented yet, this is part of https://github.com/ethersphere/go-ethereum/issues/1196 and @holisticode will take over.

When the timeout happens you have to:
1. get all references from the content (this is where you need the full content uploaded)
2. check all hosts whether they have any of the references
3. etc.

Reply:
Ok, I am sorry, then it is fine. I just reviewed the code presented here.

Reply:
Yes, it wasn't immediately clear, as this is not implemented yet. However this is the whole point of moving the `randomBytes` slice one level up in scope, so that it is available where we catch the timeout or the failure, and available for debug purposes.